lib/libzfs/libzfs_pool.c

```diff
@@ -2118,72 +2168,141 @@ zpool_initialize(zpool_handle_t *zhp, pool_initialize_func_t cmd_type,
+	nvlist_t *errlist;
 	nvpair_t *elem;
 
-	for (elem = nvlist_next_nvpair(vds, NULL); elem != NULL;
-	    elem = nvlist_next_nvpair(vds, elem)) {
-		char *vd_path = nvpair_name(elem);
-		tgt = zpool_find_vdev(zhp, vd_path, &spare, &cache, NULL);
+	err = zpool_translate_vdev_guids(zhp, vds, vdev_guids,
+	    guids_to_paths, &vd_errlist);
 
-		if ((tgt == NULL) || cache || spare) {
-			(void) snprintf(msg, sizeof (msg),
-			    dgettext(TEXT_DOMAIN, "cannot initialize '%s'"),
-			    vd_path);
-			int err = (tgt == NULL) ? EZFS_NODEVICE :
-			    (spare ? EZFS_ISSPARE : EZFS_ISL2CACHE);
+	if (err == 0) {
+		err = lzc_initialize(zhp->zpool_name, cmd_type,
+		    vdev_guids, &errlist);
+		if (err == 0) {
 			fnvlist_free(vdev_guids);
 			fnvlist_free(guids_to_paths);
-			return (zfs_error(hdl, err, msg));
+			return (0);
 		}
 
-		uint64_t guid = fnvlist_lookup_uint64(tgt, ZPOOL_CONFIG_GUID);
-		fnvlist_add_uint64(vdev_guids, vd_path, guid);
+		if (errlist != NULL) {
+			vd_errlist = fnvlist_lookup_nvlist(errlist,
+			    ZPOOL_INITIALIZE_VDEVS);
+		}
 
-		(void) snprintf(msg, sizeof (msg), "%llu", (u_longlong_t)guid);
-		fnvlist_add_string(guids_to_paths, msg, vd_path);
+		(void) snprintf(msg, sizeof (msg),
+		    dgettext(TEXT_DOMAIN, "operation failed"));
+	} else {
+		verify(vd_errlist != NULL);
+	}
+
+	for (elem = nvlist_next_nvpair(vd_errlist, NULL); elem != NULL;
+	    elem = nvlist_next_nvpair(vd_errlist, elem)) {
+		int64_t vd_error = xlate_init_err(fnvpair_value_int64(elem));
+		char *path;
+
+		if (nvlist_lookup_string(guids_to_paths, nvpair_name(elem),
+		    &path) != 0)
+			path = nvpair_name(elem);
+
+		(void) zfs_error_fmt(zhp->zpool_hdl, vd_error,
+		    "cannot initialize '%s'", path);
 	}
 
-	int err = lzc_initialize(zhp->zpool_name, cmd_type, vdev_guids,
-	    &errlist);
 	fnvlist_free(vdev_guids);
+	fnvlist_free(guids_to_paths);
 
-	if (err == 0) {
-		fnvlist_free(guids_to_paths);
-		return (0);
+	if (vd_errlist != NULL) {
+		fnvlist_free(vd_errlist);
+		return (-1);
+	}
+
+	return (zpool_standard_error(zhp->zpool_hdl, err, msg));
+}
+
+static int
+xlate_trim_err(int err)
+{
+	switch (err) {
+	case ENODEV:
+		return (EZFS_NODEVICE);
+	case EINVAL:
+	case EROFS:
+		return (EZFS_BADDEV);
+	case EBUSY:
+		return (EZFS_TRIMMING);
+	case ESRCH:
+		return (EZFS_NO_TRIM);
+	case EOPNOTSUPP:
+		return (EZFS_TRIM_NOTSUP);
 	}
+	return (err);
+}
 
+/*
+ * Begin, suspend, or cancel the TRIM (discarding of all free blocks) for
+ * the given vdevs in the given pool.
+ */
+int
+zpool_trim(zpool_handle_t *zhp, pool_trim_func_t cmd_type, nvlist_t *vds,
+    uint64_t rate, boolean_t partial)
+{
+	char msg[1024];
+	int err;
+
+	nvlist_t *vdev_guids = fnvlist_alloc();
+	nvlist_t *guids_to_paths = fnvlist_alloc();
 	nvlist_t *vd_errlist = NULL;
-	if (errlist != NULL) {
-		vd_errlist = fnvlist_lookup_nvlist(errlist,
-		    ZPOOL_INITIALIZE_VDEVS);
+	nvlist_t *errlist;
+	nvpair_t *elem;
+
+	err = zpool_translate_vdev_guids(zhp, vds, vdev_guids,
```

Comment:
This here seems to be why I have the problem of "no such device in pool". It looks like my virtual device from loopback doesn't have a label, which then can't get GUID. 

```
[root@bl-vsnap-100 vsnap]# zpool status
  pool: rpool5e2f64bc38e3475eb5395c37e1e836d8
 state: ONLINE
  scan: none requested
config:

        NAME                                   STATE     READ WRITE CKSUM
        rpool5e2f64bc38e3475eb5395c37e1e836d8  ONLINE       0     0     0
          part1                                ONLINE       0     0     0

errors: No known data errors

  pool: vpool2
 state: ONLINE
  scan: none requested
config:

        NAME                                      STATE     READ WRITE CKSUM
        vpool2                                    ONLINE       0     0     0
          scsi-36000c290f7e3d3e0c3644005631d8107  ONLINE       0     0     0

errors: No known data errors

[root@bl-vsnap-100 vsnap]# zdb -l rpool5e2f64bc38e3475eb5395c37e1e836d8
failed to find device rpool5e2f64bc38e3475eb5395c37e1e836d8, try specifying absolute path instead
```

With @dweeezil his code didn't use this he instead used zfs ioctl directly without using the vdev guids, which is why his code worked for me, but this does not.

Reply:
The other weird thing is if I put in the absolute path the label shows up:

```
[root@bl-vsnap-100 vsnap]# zdb -l /dev/bdg/7035cc52-4a75-4ced-b392-ef6bfa63fb2b/part1
------------------------------------
LABEL 0
------------------------------------
    version: 5000
    name: 'rpool5e2f64bc38e3475eb5395c37e1e836d8'
    state: 0
    txg: 94
    pool_guid: 11403916235060901602
    errata: 0
    hostname: 'bl-vsnap-100.devad.catalogic.us'
    top_guid: 6718177743574283759
    guid: 6718177743574283759
    vdev_children: 1
    vdev_tree:
        type: 'disk'
        id: 0
        guid: 6718177743574283759
        path: '/dev/bdg/7035cc52-4a75-4ced-b392-ef6bfa63fb2b/part1'
        devid: 'scsi-360014057a920dee1e9d427e86aee8a51-part1'
        whole_disk: 1
        metaslab_array: 131
        metaslab_shift: 35
        ashift: 12
        asize: 4503599538765824
        is_log: 0
        create_txg: 4
    features_for_read:
        com.delphix:hole_birth
        com.delphix:embedded_data
    labels = 0 1 2 3
```